update usb host connect disconnect

AidenHu opened the pull request:
Dependency:
 
[USB Host: integrate class API [4: aggregated] #99775](https://github.com/zephyrproject-rtos/zephyr/pull/99775)

This change is used to update usb host connection/disconnection even when usb hub is used.

Commit:
tests: usb: uvc: connect USB device and host via UVB

Add a test to run the USB Video Class host support by using
the existing Zephyr USB Video Class device support.
This allows running implementing the host side from the device side.
A draft implementation of UVC is added leveraging this test.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: uvc: move the header definition to include/

Move the UVC header with all the definitions from the UVC standard to
share it between USB host and device class implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: uvc: move the header definition to include/

Move the UVC header with all the definitions from the UVC standard to
share it between USB host and device class implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: uvc: placeholder implementation

Loop through each of the VideoStreaming and VideoControl descriptor
to parse them. This is meant as a stub for the purpose of testing the
class API.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: add host status to the host context

Add a "struct usbh_status" that contains a bitmask of flags to keep
track of the global state of the host context, like done for the
device_next implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: class: implement more of the class/data struct fields

Add a "struct usbh_class_api" for the host implementation, and move all
the function poitners to it. Add more fields to "struct usbh_class_data".

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: introduce wrappers to access the class function pointers

Add API wrappers around the function pointers in struct usbh_class_api,
while also documenting the USB host class internal API.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: introduce usbh_class with init/remove functions

Add functions to probe/remove all classes as part of a new usbh_class.c
and a matching usbh_class.h. These functions are called from the function
usbh_init_device_intl() in usbh_core.c to initialize every class upon
connection of a device. Every class driver provide filters to match the
interfaces of the device.

Co-authored-by: Aiden Hu <weiwei.hu@nxp.com>
Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
uvc: device: move helper functions to a common directory

Move UVC helper functions to a file shared between UVC host and device.
The arrays are not visible anymore from either USB host or device, but
instead accessed through a front-end funciton.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
tests: usb: host: test basic API functionality

Add tests making sure the USB Host class APIs introduced build
and run as expected.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
tests: usb: host: test basic API functionality

Add tests making sure the USB Host class APIs introduced build
and run as expected.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: add copyright notice to linker script

Add missing copyright notice for the linker script to help with
check_compliance.py.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: device_next: uvc: fix public API include

The public API file <zephyr/usb/class/usbd_uvc.h> lacked an include
to <zephyr/drivers/video.h> making it fail depending on the order of
the includes.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

santhosh-c-c committed:
usb: host: device: Read descriptor after setting device address

Modify the USB device int sequence to read the device
descriptor only after setting a valid device address.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>

AidenHu committed:
drivers: uhc: use correct endpoint type and interval

For usb xfer, set endpoint type and interval by the
selected endpoint desc.

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

AidenHu committed:
drivers: uhc: correct interval usage for usb xfer

Convert xfer's interval to actual value because
mcux_ep->interval is already calculated.

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

AidenHu committed:
subsys: usb: host: support multiple devices under hub

When hub is used, need to consider about multiple
devices are attached.

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

AidenHu committed:
drivers: uhc: set right value for pipe by xfer's mps

maxPacketSize and numberPerUframe of pipe should
be set considering additional transactions.

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

AidenHu committed:
subsys: usb: host: remove root member for usbh_context

add usbh_device_get_root and usbh_device_is_root
function to check root device

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

AidenHu committed:
subsys: usb: host: refine connect/disconnect handling

Add two functions:
usbh_connect_device() for device connection
usbh_disconnect_device() for device disconnection

These functions centralize the logic for device attach/detach,
including class probe and remove handling. They can be invoked
by the hub class as well as dev_connected_handler and
dev_removed_handler, improving code clarity and reuse.

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

AidenHu commented:
@josuah 
This is created by [API4](https://github.com/zephyrproject-rtos/zephyr/pull/99775) and I have tested based on the newest [hub_class ](https://github.com/zephyrproject-rtos/zephyr/pull/99735/commits/4cf9c8c6dd743d57d7c3571f881a453cc1d44048)for multiple devices. Hope you can review.

josuah commented on subsys/usb/host/usbh_device.c:
```diff
@@ -475,6 +477,58 @@ bool usbh_device_is_root(struct usbh_context *const ctx,
+		LOG_ERR("Failed to init new USB device");
+		k_mutex_lock(&ctx->mutex, K_FOREVER);
+		sys_dlist_remove(&udev->node);
+		k_mutex_unlock(&ctx->mutex);
```
Likewise, `usbh_device_free()` takes care of calling `sys_dlis_remove();`, so not needed here (unless it is removed from `usbh_device_free()`.

AidenHu commented on subsys/usb/host/usbh_device.c:
```diff
@@ -475,6 +477,58 @@ bool usbh_device_is_root(struct usbh_context *const ctx,
+
+	k_mutex_lock(&ctx->mutex, K_FOREVER);
+	sys_dlist_append(&ctx->udevs, &udev->node);
+	k_mutex_unlock(&ctx->mutex);
```
Yes, it is. Remove it from `usbh_connect_device()`

AidenHu commented:
> Once comments above adjusted, I can integrate it to API4, and if that makes sense to you, it is possible to put this PR on top of `main` directly, with just the last commit, then it can be reviewed+merged without waiting..

@josuah 
I have updated this commit based your comments, thank you much. Meanwhile, I also rebase the top of main for the current PR [update usb host connect disconnect#100072](https://github.com/zephyrproject-rtos/zephyr/pull/100072), which means that  [update usb host connect disconnect#100072](https://github.com/zephyrproject-rtos/zephyr/pull/100072) can be the last pull request for reviewing and merged to be main branch as what we have talked in Discord.
If something wrong I did, please correct me, thanks.

Commit:
usb: uvc: common: correct the size for UVC_PU_CONTRAST_CONTROL

The USB control size field was wrong for UVC_PU_CONTRAST_CONTROL.
Correct it to the correct value from the standard.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>
